### PR TITLE
[WIP] Add testing against Core (excl USB)

### PR DESCRIPTION
**Pull request by CoMPaTech**
This - for now draft PR - aims at (as) fully (as possible) testing the `custom_component` against Core, as such the actions will perform this by replacing our integration in Core with the custom_component parts. It will (purposely!) **not** include USB coverage so we can easily diverge between this PR and an upcoming one for upstreaming later.

 - [ ] tests/fixtures/plugwise - replaced by the test-contents of this repo (should be backfilled from `python-plugwise` obviously
 - [ ] tests/components/plugwise - replaced by the test-contents of this repo
 - [ ] homeassistant/components/plugwise - replaced by the custom_component of this repo (forgoing translations which can not be PRed eventually)

At time of writing code is eventually `pytest`ed first before `flake8/pylint` (and then `pytest`ed again). That first `pytest` should be removed from the action to better mimick the flow of core PRs - which also means that at that point in time we should get `pre-commit` working like we have in `python-plugwise`.

### CODE REVIEW

**Comment by CoMPaTech** [comment]
Scaffolding mostly done - with some flake8/pylint fixes already in place - apparently (or todo) the `tests/` code isn't linted in HA-Core so skipped that for now. 

Todo from here is making the `tests` work (and most probably also upload newer `fixtures`) with our upgraded module-code.

**Comment by CoMPaTech** [comment]
Small progress on testing, config_flow is mostly 'fixed' in terms of working again (there is however a `TODO` now) aside from further tests being required on new stuff and the USB part. For now getting the tests to work is the most important factor, from there we can work our way up to coverage again.
CodeFactor will popup the use of passwords again, as we probably won't be able to 'string' and 'random' on Core I'm fine with leaving that problem in and/or ignoring it.

**Comment by CoMPaTech** [comment]
Aside from USB-stuff (which should be added later) there are a few things left todo in `test_config_flow.py` which are noted in a todolist on the top for now. Most buggy one is where the `already_configured` isn't working anymore - the others are newly added code that needs tests.

**Review by CoMPaTech — COMMENTED** [review]
Check

**Comment by CoMPaTech** [comment]
Regarding 8cc4e73 -> when doing either 'user-based-config' I ran into the problem with not aborting. This is due to `unique_id` being `None` (and the upstream abort doesn't trigger on None apparently).

Testing on username and then zeroconf seemed out of line, so I just do 2x zeroconf know (although that shouldn't really matter) and unique_id is **only** set during the user_step, not beforehand in zeroconf (as zeroconf calls user_step). This irons out the 'abort if configured' not being tested.

**Comment by CoMPaTech** [comment]
Summary:
`homeassistant/components/plugwise/config_flow.py       174     66    62%   70-74, 79-102, 107-113, 195-224, 234-246, 304-305, 329-333, 338`

Double-checked L70/79/107/195/234 are USB (for @brefra skipping this for now to make a clean cut between 'what needs upstreaming' and 'what was partially/already in `plugwise-usb`')

L329 and L338: Will discuss with @bouwew on Discord

File: custom_components/plugwise/config_flow.py [hunk 1]
```diff
@@ -285,7 +281,7 @@ async def async_step_user_gateway(self, user_input=None):
 
             if not errors:
                 await self.async_set_unique_id(
-                    api.smile_hostname or api.gateway_id, raise_on_progress=False
+                    api.smile_hostname or api.gateway_id or user_input[CONF_HOST].split(".")[0], raise_on_progress=False
```
Comment (bouwew) [inline custom_components/plugwise/config_flow.py]: This added bit is not needed? The unique_id should either be the smile hostsname or the gateway_id. \We've determined in the past that these two option were sufficient to cover all known Smiles and Stretches.

Comment (CoMPaTech) [reply custom_components/plugwise/config_flow.py]: I'll investigate then, because without it and using zeroconf there was no api.smile_hostname :(

Comment (CoMPaTech) [reply custom_components/plugwise/config_flow.py]: So we should patch elsewhere then ... removing that part breaks the 'already_configured'.

Comment (bouwew) [reply custom_components/plugwise/config_flow.py]: api.smile_hostname needs `connect()` to be called.

Comment (CoMPaTech) [reply custom_components/plugwise/config_flow.py]: Yeah, which is where it falls apart, that just returns `True` now, but obviously doesn't set the required parameters. I'll dig a little deeper later

Comment (CoMPaTech) [reply custom_components/plugwise/config_flow.py]: Done

**Comment by CoMPaTech** [comment]
For the rest of our coverage (leaving out usb-parts) I only see minors and nothing major or urgent. 

For future reference:
 - Gateway L82-83 and L90-92 might be a nice addition at some point.
 - Sensor L172's new heating_state might be a nice addition (will need fixtures I guess)

I'll set it to PR instead of just draft - I'll have to recheck the steps (especially changes to `custom_components/plugwise` as those must be ok with us all)

**Review by bouwew — APPROVED** [review]
Looking very good!
I haven't looked at the testing code in detail, this will take me much more time :)

**Pull request merged**